src/vaults/restaking/WithdrawalEscrowV2.sol

```diff
@@ -0,0 +1,185 @@
+
+    struct EpochInfo {
+        uint128 shares;
+        uint128 assets;
+    }
+
+    uint256 public currentEpoch;
+    uint256 public resolvingEpoch;
+    uint256 public totalDebt;
+
+    // map per epoch debt share
+    mapping(uint256 => EpochInfo) public epochInfo;
+
+    constructor(UltraLRT _vault) {
+        asset = ERC20(_vault.asset());
+        vault = _vault;
+    }
+
+    modifier onlyVault() {
+        require(msg.sender == address(vault), "WE: must be vault");
+        _;
+    }
+
+    modifier onlyGovernance() {
+        require(msg.sender == address(vault.governance()), "WE: Must be gov");
+        _;
+    }
+
+    /**
+     * @notice Withdrawal Request event
+     * @param user user address
+     * @param epoch epoch of the request
+     * @param shares withdrawal vault shares
+     * @dev will makes things easy to search for each user withdrawal requests
+     */
+    event WithdrawalRequest(address indexed user, uint256 epoch, uint256 shares);
+
+    /**
+     * @notice Register withdrawal request as debt
+     * @param user user address
+     * @param shares amount of vault shares user requested to withdraw
+     */
+    function registerWithdrawalRequest(address user, uint256 shares) external onlyVault {
+        // register shares of the user
+
+        userDebtShare[currentEpoch][user] += shares;
+        epochInfo[currentEpoch].shares += uint128(shares);
+
+        totalDebt += shares;
+
+        emit WithdrawalRequest(user, currentEpoch, shares);
+    }
+
+    function endEpoch() external onlyVault {
+        require(epochInfo[currentEpoch].shares > 0, "WEV2: No Debt.");
+
+        currentEpoch = currentEpoch + 1;
+        // TODO: epoch end event
+    }
+
+    function getDebtToResolve() external view returns (uint256) {
+        return resolvingEpoch < currentEpoch ? epochInfo[resolvingEpoch].shares : 0;
+    }
+
+    /**
+     * @notice resolve the locked shares for current epoch
+     * @dev This function will be triggered after closing a position
+     * @dev will check for available shares to burn
+     * @dev after resolving vault will send the assets to escrow and burn the share
+     */
+    function resolveDebtShares() external onlyVault {
+        require(resolvingEpoch < currentEpoch, "WEV2: No debt.");
+
+        uint256 preAssets = asset.balanceOf(address(this));
+        vault.redeem({shares: epochInfo[resolvingEpoch].shares, receiver: address(this), owner: address(this)});
+
+        uint256 currentAssets = asset.balanceOf(address(this));
+        totalDebt -= epochInfo[resolvingEpoch].shares;
+        epochInfo[resolvingEpoch].assets = uint128(currentAssets - preAssets);
+
+        resolvingEpoch += 1;
+    }
+
+    /**
+     * @notice Redeem withdrawal request
+     * @param user address
+     * @param epoch withdrawal request epoch
+     * @return received assets
+     */
+    function redeem(address user, uint256 epoch) external returns (uint256) {
```

Comment:
maybe allow 1 tx for redeeming all user's pending withdrawals

Reply:
added `redeemMultiEpoch` function to redeem multiple epoch in single tx.